rasa/core/policies/rule_policy.py

```diff
@@ -270,6 +277,137 @@ def _check_rule_restriction(
+            featurized_slots
+        )
+        current_slots = set(state.get(SLOTS, {}).keys())
+        if expected_slots == current_slots:
+            # all expected slots are satisfied
+            return set()
+
+        return expected_slots
+
+    @staticmethod
+    def _check_active_loops_fingerprint(
+        fingerprint: Dict[Text, List[Text]], state: State
+    ) -> Set[Text]:
+        expected_active_loops = set(fingerprint.get(ACTION_FINGERPRINT_ACTIVE_LOOPS))
+        # we don't use tracker.active_loop_name
+        # because we need to keep should_not_be_set
+        current_active_loop = state.get(ACTIVE_LOOP, {}).get(LOOP_NAME)
+        if current_active_loop in expected_active_loops:
+            # one of expected active loops is set
+            return set()
+
+        return expected_active_loops
+
+    @staticmethod
+    def _error_messages_from_fingerprints(
+        action_name: Text,
+        fingerprint_slots: Set[Text],
+        fingerprint_active_loops: Set[Text],
+        rule_name: Text,
+    ) -> List[Text]:
+        error_messages = []
+        if action_name and fingerprint_slots:
+            error_messages.append(
+                f"- the action '{action_name}' in rule '{rule_name}' does not set "
+                f"slots, that it sets in other rules: "
+                f"'{', '.join(fingerprint_slots)}'. Please update the rule with "
+                f"an appropriate slot or if it is the last action add "
+                f"`wait_for_user_input: false` after this action."
+            )
+        if action_name and fingerprint_active_loops:
+            # substitute `SHOULD_NOT_BE_SET` with `null` so that users
+            # know what to put in their rules
+            fingerprint_active_loops = set(
+                "null" if active_loop == SHOULD_NOT_BE_SET else active_loop
+                for active_loop in fingerprint_active_loops
+            )
+            # add action_name to active loop so that users
+            # know what to put in their rules
+            fingerprint_active_loops.add(action_name)
+
+            error_messages.append(
+                f"- the action '{action_name}' in rule '{rule_name}' doesn't set an "
+                f"active_loop, that it sets in other rules: "
+                f"'{', '.join(fingerprint_active_loops)}'. Please update the rule with "
+                f"an appropriate active loop or if it is the last action add "
+                f"`wait_for_user_input: false` after this action."
+            )
+        return error_messages
+
+    def _check_rule_action_fingerprints(
+        self, rule_trackers: List[TrackerWithCachedStates], domain: Domain,
+    ) -> None:
+        import rasa.core.policies.ensemble
+
+        logger.debug("Started checking if some rules are incomplete.")
+        # we need to use only fingerprints from rules
+        rule_fingerprints = rasa.core.policies.ensemble.create_action_fingerprints(
```

Comment:
What was the reasoning behind putting these functions into to `ensemble`? To me this seems to rather fit into the `tracker` module

Reply:
have no idea. this are the old functions that were in ensemble class

Reply:
I just found the new usage for them

Reply:
moved to `trackers.py`